Test users

ErwinWoz opened the pull request:
please take a closer look at TestReadAllNames and TestFilterNames, they can potentially be improved or redesigned

ErwinWoz committed:
added first batch of user tests

ErwinWoz committed:
added org and role tests

dprinzing committed:
source code formatting tweaks

dprinzing committed:
clean up lint issue

dprinzing commented on pkg/user/users_test.go:
```diff
@@ -0,0 +1,606 @@
+	expect := assert.New(t)
+	allVersions, err := service.ReadAllVersions(ctx, id2)
+	if expect.NoError(err) {
+		expect.Equal(u20, allVersions[0])
```
Note that `ReadAllVersions` can return an empty slice, without error. We should verify that `allVersions` is not empty.

dprinzing commented on pkg/user/users_test.go:
```diff
@@ -0,0 +1,606 @@
+		onlyNames := v.Map(idsAndNames, func(entry v.TextValue) string {
+			return entry.Value
+		})
+		expect.Equal(len(onlyNames), len(knownIDs))
```
`onlyNames` might contain the name of the transient CRUD user, so we should use `GreaterOrEqual`.

dprinzing commented on pkg/user/users_test.go:
```diff
@@ -0,0 +1,606 @@
+func TestReadAllUsersByOrgID(t *testing.T) {
+	expect := assert.New(t)
+	users, err := service.ReadAllUsersByOrgID(ctx, orgID1)
+	if expect.NoError(err) && expect.NotEmpty(users) {
```
Since there's only one user in Organization 1, we could simplify this to verifying that the length of `users` is 1 and it contains `u11`.

dprinzing commented on pkg/user/users_test.go:
```diff
@@ -0,0 +1,606 @@
+func TestReadAllUsersByRole(t *testing.T) {
+	expect := assert.New(t)
+	usersByRole, err := service.ReadAllUsersByRole(ctx, "assistant")
+	if expect.NoError(err) && expect.NotEmpty(usersByRole) {
```
The CRUD user does not have the role "assistant", so we expect the number of users to be 4. This block could be simplified to testing that the length of `usersByRole` is 4.

dprinzing commented on pkg/user/users_test.go:
```diff
@@ -0,0 +1,606 @@
+			numOfUsers = append(numOfUsers, i)
+			usersIDs = append(usersIDs, v.ID)
+		}
+		expect.GreaterOrEqual(len(numOfUsers), 4)
```
This could be simplified to testing the length of `checkUsers`; we don't need `numOfUsers`.

dprinzing commented on pkg/user/users_test.go:
```diff
@@ -0,0 +1,606 @@
+			usersIDs = append(usersIDs, v.ID)
+		}
+		expect.GreaterOrEqual(len(numOfUsers), 4)
+		expect.Subset(knownIDs, usersIDs)
```
I think these two arguments should be reversed. `userIDs` might contain the ID of the CRUD user.

dprinzing committed:
Updated comments

dprinzing reviewed (changes requested):
A few small comments. Otherwise, this looks great!

ErwinWoz committed:
updated tests based on PR comments

ErwinWoz committed:
edited userTestReadAllUsersByStatus

ErwinWoz committed:
merge changes

dprinzing reviewed (approved):
Looks good! I think it's ready to merge.

ErwinWoz commented:
Merged branch test-users into main and deleted the branch.

Added test file for pkg/users directory

Thanks to @dprinzing for conducting the review process
